### PR TITLE
sd-ndisc: do not stop solicitations on zero-lifetime RA

### LINKED ISSUE #43205: sd-ndisc: any received RA cancels the pending initial Router Solicitation, even a lifetime-0 RA that networkd itself then discards
### systemd version the issue has been seen with

255.4-1ubuntu8.16 — code re-checked on `main`, still present

### Used distribution

Ubuntu 24.04 (noble)

### Linux kernel version used

6.8.0-136-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd, sd-ndisc

### Expected behaviour you didn't see

Per RFC 4861 §6.3.7, a host desists from sending solicitations only *"Once the host sends a Router Solicitation, and receives a valid Router Advertisement with a **non-zero** Router Lifetime"*.

An RA with Router Lifetime 0 and no prefix information — in particular one the host emitted itself, which networkd goes on to explicitly discard — should not cancel an initial RS that has not been sent yet.

### Unexpected behaviour you saw

`ndisc_handle_router()` disables both timers with no guard whatsoever:

```c
(void) event_source_disable(nd->timeout_event_source);
(void) event_source_disable(nd->timeout_no_ra);
```

No check on Router Lifetime, on source address, or on whether an RS was ever sent.

Consequence: if **any** RA lands inside the `MAX_RTR_SOLICITATION_DELAY` random window, the initial RS is never transmitted. When that RA carries nothing usable, networkd has neither solicited nor learned anything, and must wait for the next *unsolicited* RA — up to `MaxRtrAdvInterval`, which is 600 s on our upstream (odhcpd) and may be 1800 s per RFC 4861.

There is an internal inconsistency here: the very same RA is rejected a moment later by networkd —

```
enp1s0: No NDisc route added, gateway fe80::…:2b08 matches local address
```

— yet it has already been used to cancel the solicitation.

### Steps to reproduce the problem

The host is an IPv6 router advertising a DNS-only RA (Router Lifetime 0, RDNSS only) on the same interface networkd manages for the uplink.

1. `systemctl restart systemd-networkd`
2. within ~10 ms, send a minimal RA (type 134, Router Lifetime 0, no options) to `ff02::1` from this host's own link-local address
3. no RS is ever emitted; the SLAAC address stays gone until the upstream's next unsolicited RA

Isolated with a **single-bit control** — identical in every other respect, only `IPV6_MULTICAST_LOOP` on the injecting socket differs (the RA goes out on the wire either way; the bit only decides whether it loops back to networkd's ndisc socket on the same host):

| `IPV6_MULTICAST_LOOP` | RA reaches networkd | RS emitted | SLAAC address |
|---|---|---|---|
| 1 | yes | **none in 60 s** | never returns |
| 0 | no | **+0.83 s** | back at +1.11 s |

### Additional program output to the terminal or log subsystem illustrating the issue

Working path (no RA inside the window):

```
09:31:55.971938 enp1s0: Requested configuring of the IPv6 Router Discovery.
09:31:57.231765 wire: fe80::…:2b08 > ff02::2: ICMP6, router solicitation
09:31:57.232489 wire: fe80::2e0:… > fe80::…:2b08: ICMP6, router advertisement
```

Broken path (RA at +6.6 ms):

```
09:35:38.132241 enp1s0: NDISC: Started IPv6 Router Solicitation client
09:35:38.132245 enp1s0: IPv6 Router Discovery is configured and started.
09:35:38.132562 enp1s0: State changed: configuring -> configured
09:35:38.138882 enp1s0: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
09:35:38.139032 enp1s0: No NDisc route added, gateway fe80::…:2b08 matches local address
09:35:38.639951 enp1s0: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
(no RS on the wire; global address absent for 90 s until forced with `rdisc6 -1`)
```

Suggested fix: gate the cancellation on the RFC 4861 §6.3.7 conditions — an RS having been sent *and* the RA carrying a non-zero Router Lifetime. At minimum, an RA whose source is one of the host's own addresses (already recognised as such downstream) should not suppress solicitation.

### PR DESCRIPTION
**Pull request by quanyeyang**
RFC 4861 section 6.3.7 says that a host should send at least one Router Solicitation when an advertisement is received before the first solicitation, and should desist only after receiving a valid Router Advertisement with a non-zero Router Lifetime.

Keep the RS timer active until both conditions are met, and add a regression test for a lifetime-zero RA arriving before the initial RS.

Fixes #43205.

### CODE REVIEW

File: src/libsystemd-network/test-ndisc-rs.c [hunk 1]
```diff
@@ -295,6 +295,15 @@ static void test_callback_count(
                 (*count)++;
 }
 
+static int on_recv_rs_exit(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        _cleanup_(icmp6_packet_unrefp) ICMP6Packet *packet = NULL;
+
+        assert_se(icmp6_packet_receive(fd, &packet) >= 0);
```
Comment (poettering) [inline src/libsystemd-network/test-ndisc-rs.c]: new style test macros?

File: src/libsystemd-network/sd-ndisc.c [hunk 2]
```diff
@@ -219,7 +219,12 @@ static int ndisc_handle_router(sd_ndisc *nd, ICMP6Packet *packet) {
         if (r < 0)
                 return r;
 
-        (void) event_source_disable(nd->timeout_event_source);
+        r = sd_ndisc_router_get_lifetime(rt, NULL);
+        if (r < 0)
+                return r;
+
+        if (nd->retransmit_time > 0 && r > 0)
```
Comment (yuwata) [inline src/libsystemd-network/sd-ndisc.c]: The first condition `nd->retransmit_time > 0` is redundant?
```c
if (r > 0)
```
should be sufficient.

File: src/libsystemd-network/test-ndisc-rs.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd-network/test-ndisc-rs.c]: Yeah, please use `ASSERT_OK()` here, and use similar macros in the new test case.

File: src/libsystemd-network/sd-ndisc.c [hunk 3]
```diff
@@ -219,7 +219,12 @@ static int ndisc_handle_router(sd_ndisc *nd, ICMP6Packet *packet) {
         if (r < 0)
                 return r;
 
-        (void) event_source_disable(nd->timeout_event_source);
+        r = sd_ndisc_router_get_lifetime(rt, NULL);
+        if (r < 0)
+                return r;
+
+        if (r > 0)
```
[the remark this answers is not in the document]
Comment (yuwata) [reply src/libsystemd-network/sd-ndisc.c]: fixed

File: src/libsystemd-network/test-ndisc-rs.c [hunk 4]
```diff
@@ -323,6 +332,39 @@ TEST(rs_ifindex_mismatch) {
...
+        ASSERT_OK(sd_ndisc_set_callback(nd, test_callback_count, &count));
+        ASSERT_OK(sd_event_add_time_relative(e, NULL, CLOCK_BOOTTIME,
+                                             30 * USEC_PER_SEC, 0,
+                                             NULL, INT_TO_PTR(-ETIMEDOUT)));
+        ASSERT_OK(sd_ndisc_start(nd));
+
+        ASSERT_EQ(write(test_fd[1], &advertisement, sizeof(advertisement)), (ssize_t) sizeof(advertisement));
+        ASSERT_OK_POSITIVE(sd_event_run(e, UINT64_MAX));
+        ASSERT_EQ(count, 1U);
+        ASSERT_OK_POSITIVE(sd_event_source_get_enabled(nd->timeout_event_source, NULL));
```
[the remark this answers is not in the document]
Comment (yuwata) [reply src/libsystemd-network/test-ndisc-rs.c:358]: No thank you.

File: src/libsystemd-network/sd-ndisc.c [hunk 5]
```diff
@@ -219,7 +219,12 @@ static int ndisc_handle_router(sd_ndisc *nd, ICMP6Packet *packet) {
...
+                (void) event_source_disable(nd->timeout_event_source);
```
[the remark this answers is not in the document]
Comment (yuwata) [reply src/libsystemd-network/sd-ndisc.c:224]: fixed

**Comment by yuwata** [comment]
@quanyeyang Thank you! I slightly adjusted the commits to address suggestions by Claude and squash them.
